source/Halibut.Tests/ExceptionContractFixture.cs

```diff
@@ -0,0 +1,290 @@
+
+            var client = clientAndService.CreateAsyncClient<IEchoService, IAsyncClientEchoServiceWithOptions>();
+
+            (await AssertException.Throws<HalibutClientException>(async () => await client.SayHelloAsync("Hello", new(CancellationToken, CancellationToken.None))))
+                .And.Message.Should().ContainAny(
+                    $"An error occurred when sending a request to '{clientAndService.ServiceUri}', before the request could begin: Unable to read data from the transport connection:",
+                    $"An error occurred when sending a request to '{clientAndService.ServiceUri}', before the request could begin: Unable to write data to the transport connection:");
+        }
+        
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testNetworkConditions: false, testPolling: false, testWebSocket: false)]
+        public async Task WhenTheListeningRequestFailsToBeSent_AsTheConnectionAttemptToTheServiceTimesOut_AHalibutClientExceptionShouldBeThrown(ClientAndServiceTestCase clientAndServiceTestCase)
+        {
+            var halibutTimeoutsAndLimits = new HalibutTimeoutsAndLimitsForTestsBuilder().Build();
+
+            halibutTimeoutsAndLimits.TcpClientConnectTimeout = TimeSpan.FromSeconds(1);
+
+            await using var clientOnly = await clientAndServiceTestCase.CreateClientOnlyTestCaseBuilder()
+                .AsLatestClientBuilder()
+                .WithHalibutTimeoutsAndLimits(halibutTimeoutsAndLimits)
+                .Build(CancellationToken);
+
+            // We need to use a non localhost address to get a timeout
+            var client = clientOnly.CreateClient<IEchoService, IAsyncClientEchoServiceWithOptions>(new Uri("https://20.5.79.31:10933/"));
+
+            (await AssertException.Throws<HalibutClientException>(async () => await client.SayHelloAsync("Hello", new(CancellationToken, CancellationToken.None))))
+                .And.Message.Should().Contain(
+                    $"An error occurred when sending a request to 'https://20.5.79.31:10933/', before the request could begin: The client was unable to establish the initial connection within the timeout 00:00:01.");
+        }
+
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testNetworkConditions: false, testPolling: false, testWebSocket: false)]
+        public async Task WhenTheListeningRequestIsCancelledImmediatelyWhileTryingToConnect_AHalibutClientExceptionShouldBeThrown(ClientAndServiceTestCase clientAndServiceTestCase)
+        {
+            var halibutTimeoutsAndLimits = new HalibutTimeoutsAndLimitsForTestsBuilder().Build();
+            halibutTimeoutsAndLimits.TcpClientConnectTimeout = TimeSpan.FromSeconds(5);
+            halibutTimeoutsAndLimits.ConnectionErrorRetryTimeout = TimeSpan.FromSeconds(600);
+            halibutTimeoutsAndLimits.RetryCountLimit = int.MaxValue;
+
+            var connectionsObserver = new TestConnectionsObserver();
```

Comment:
What's the observer used for?

Reply:
Occupying a small amount of memory 😉  Good spot 🙇

Reply:
Removed